### PR TITLE
Make NameError#missing_name work even for real Ruby NameError

### PR DESCRIPTION
**Pull request by casperisfine**
As discussed in https://github.com/rails/rails/pull/37632

### Context

On constant missing Ruby call `#inspect` on the receiver to build the error message.

For instance, the error message for `Foo::Bar` will be `"#{Foo.inspect}::Bar"`.

And since Active Record override the model classes inspect method, this breaks `missing_name` assumptions.

Until now it worked because missing_name was only called on errors raised by the classic autoloader, and the classic autoloader calls `#name` to build its error message.

### Solution

`NameError` provides both `#receiver` and `#name`, so we can actually build the expected constant name without relying on a regexp. So I simply use that here.

However that PR isn't quite finalized yet, I need to:

  - Check that these methods are available on all supported rubies.
  - Check that the classic autoloader properly set these two properties.

@fxn 

cc @rafaelfranca @etiennebarrie @Edouard-chin

### CODE REVIEW

**Comment by casperisfine** [comment]
Ok, so according to CI, it looks like `receiver` is indeed not available on Ruby 2.5.

I'll figure some fallback.

**Comment by casperisfine** [comment]
Scratch that, it's available, it's just that it raises ArgumentError if it's not set, and there is no setter for it... But I have a few ideas how to deal with that.

File: actionpack/test/controller/helper_test.rb [hunk 1]
```diff
@@ -88,7 +88,7 @@ class HelpersTypoControllerTest < ActiveSupport::TestCase
   def test_helper_typo_error_message
     e = assert_raise(NameError) { HelpersTypoController.helper "admin/users" }
     # This message is better if autoloading.
-    assert_equal "uninitialized constant Admin::UsersHelper", e.message
+    assert_equal "uninitialized constant Admin::UsersHelper\nDid you mean?  Admin::UsersHelpeR", e.message
```
Comment (casperisfine) [inline actionpack/test/controller/helper_test.rb:91]: Setting `receiver` make did_you_mean work. It's probably a good thing.

File: activesupport/lib/active_support/core_ext/name_error.rb [hunk 2]
```diff
@@ -1,6 +1,13 @@
 # frozen_string_literal: true
 
 class NameError
+  attr_writer :receiver
+  alias_method :original_receiver, :receiver
+
+  def receiver
+    @receiver ||= original_receiver
```
Comment (casperisfine) [inline activesupport/lib/active_support/core_ext/name_error.rb:8]: So MRI store the `receiver` in a private instance variable (`receiver` with no `@`), there is no way to set it from Ruby-land.

**Review by casperisfine — COMMENTED** [review]
@fxn I think i need your opinion on this. I think this PR works, but I'm not certain it's the direction we want to go.

However if we decide to go that way, then maybe `Zeitwerk::NameError` should also expose `receiver`.

File: activesupport/lib/active_support/core_ext/name_error.rb [hunk 3]
```diff
@@ -1,6 +1,13 @@
 # frozen_string_literal: true
 
 class NameError
+  attr_writer :receiver
+  alias_method :original_receiver, :receiver
```
Comment (etiennebarrie) [inline activesupport/lib/active_support/core_ext/name_error.rb:5]: How about prepending a module to avoid adding `original_receiver` to `NameError`'s API?

File: activesupport/lib/active_support/core_ext/name_error.rb [hunk 4]
```diff
@@ -1,6 +1,13 @@
 # frozen_string_literal: true
 
 class NameError
+  attr_writer :receiver
```
Comment (etiennebarrie) [inline activesupport/lib/active_support/core_ext/name_error.rb:4]: I think we may want to `:nodoc:` this, unless you see some other use that people could have with it?

**Comment by casperisfine** [comment]
So turns out I missed that from 2.6 onwards, `NameError` accepts a `receiver:` keyword parameter to it's constructor.

So I should restrict that shim to 2.5

**Comment by casperisfine** [comment]
All green now.

**Comment by fxn** [comment]
Hey, it's being a busy day, I'll check it out soon. Also have in the radar the issue in the Ruby tracker.

**Comment by casperisfine** [comment]
No worries, I monkey patched it for now.

**Comment by fxn** [comment]
While `missing_name?` is public API, Rails uses `missing_name?` in a couple of places only.

I think I would prefer a simpler solution that does not reopen `NameError`. Also, Zeitwerk right now has support back to 2.4.x, and while if available it would be nice to set the receiver (I'll do it), I would prefer not to monkey patch such a core class.

If Ruby calls `inspect` on the parent module to build the error message, then what is happening is that `missing_name?` is currently based on a wrong assumption. Either the implementation needs to get the parent module object from `name` via `constantize`, or the method accepts that the caller passes a receiver.

What if we write something optimistic like:

1. If the exception has a receiver, use it.
2. Otherwise, try optimistically `name` as we do today.
3. Otherwise, obtain the parent dynamically and build the string.

What do you think?

**Comment by casperisfine** [comment]
>  it would be nice to set the receiver (I'll do it)

Actually I checked, and Zeitwerk never raise `uninitialized constant` `NameError`. So there's no change needed for Zeitwerk.

> What do you think?

I think it's more complicated, and less reliable than my PR. As long as `AS::Dependencies` sets the `receiver` properly, we no longer have to rely on error message parsing.

**Comment by casperisfine** [comment]
So I figured why we don't get the actual `inspect` return, it's a very old Ruby limitation: https://github.com/ruby/ruby/pull/3090

**Comment by fxn** [comment]
Hey Jean!

It's late here, but let me follow up with some thoughts.

We both agree that if we have a `receiver` and a `name`, that's the most robust solution.

For example, in order to make Rails itself robust here, we could define a new utility method outside `NameError` that receives the exception object, the receiver, and the constant name. We control the call sites, there are only two, and in both of them (if my tired neurons read the code correctly), we have the receiver and the expected constant name and can look for the expected string in the message. I would also prefer not to do a lookup in the error message, but we have to play the hand of cards that we got. More on that later.

Zeitwerk may raise `Zeitwerk::NameError` if the file for the default controller helper exists and defines the wrong constant (you know that, just setting up the argument). However, the error message does not belong to the public interface and we cannot program the logic according to what the error message says or does not say. If the file exists and the constant is not found, we want to reraise, but checking for the exception class when looking for the default helper would be the way to go. We can even load Z code unconditionally for 6.1 to make sure the error class is defined.

On the other hand, if the application runs in `zeitwerk` mode, while autoloading there is a regular `NameError`, and Ruby is 2.5, if I am not mistaken there is no receiver because that one is not generated by AS::Dependencies.

Bottom line is, we need to juggle a combination of things and take some compromise:

* `missing_name?` is public API, do we want to fix the method in all its genericity? or do we want to fix Rails usage only?
* We need to support Ruby < 2.6.
* We need to support `classic` and `zeitwerk` modes.

There is compromise hidden in those conditions, and intuitively I don't feel this approach fits in the right spot. (Though the detailed work of understanding the implementation, what happens in different Ruby versions, and the Ruby ticket, have been awesome and worthwhile by themselves.)

Let me sleep over it, and I'll see which patch would I write, so we can compare concrete solutions and the trade-offs taken in each option.

❤️

**Comment by byroot** [comment]
> On the other hand, if the application runs in zeitwerk mode, while autoloading there is a regular NameError, and Ruby is 2.5, if I am not mistaken there is no receiver because that one is not generated by AS::Dependencies..

I think that's where the misunderstanding is. Ruby 2.5 does provide the `receiver` on real name errors:

```
>> RUBY_VERSION
=> "2.5.1"
>> Kernel::FOO rescue $!
=> #<NameError: uninitialized constant Kernel::FOO>
>> _.receiver
=> Kernel
```

Also it's documented: https://ruby-doc.org/core-2.5.0/NameError.html#method-i-receiver


What it doesn't allow it to pass a receiver when instantiating `NameError`: `NameError.new("msg", )`:

```
>> NameError.new("uninitialized constant Kernel::FOO", :Foo, receiver: Kernel)
ArgumentError (wrong number of arguments (given 2, expected 0..1))
```

So that's actually the `NameError` generated by `ActiveSupport::Dependencies` that don't have a receiver.

#### Summary

Trying to step back and better summarize here.

We have two possible method to extract the class path:

We match the error message (current rails), but that doesn't work with some real `NameError` thrown by Ruby itself, so only reliably work with Classic.

Or we can use the `receiver` and `name` attributes of the exception, but then Classic can't set them on Ruby 2.5 (unless we backport the constructor from 2.6).

**Comment by casperisfine** [comment]
Here's an alternative version that doesn't backport the `receiver:` argument for 2.5: https://github.com/Shopify/rails/commit/7ab9d9d5822fff219bd737bb3c675061d0594412

**Comment by fxn** [comment]
> I think that's where the misunderstanding is. Ruby 2.5 does provide the receiver on real name errors

Aaaahhh, I misunderstood this.

> Here's an alternative version that doesn't backport the receiver: argument for 2.5: Shopify@7ab9d9d

This patch is great for my taste.

**Comment by byroot** [comment]
Awesome!

**Comment by casperisfine** [comment]
I updated the PR to be https://github.com/Shopify/rails/commit/7ab9d9d5822fff219bd737bb3c675061d0594412 plus a few fixes for the test suite.

**Comment by fxn** [comment]
❤️
